### PR TITLE
ISPN-8428 Add TX tests for off-heap

### PR DESCRIPTION
**Pull request by vjuranek**
https://issues.jboss.org/browse/ISPN-8428

### CODE REVIEW

**Comment by wburns** [comment]
There are some checkstyle errors.

**Comment by vjuranek** [comment]
Fix, but please don't merge yet, there's a discussion under [ISPN-8429](https://issues.jboss.org/browse/ISPN-8429), if it's a bug or not (if not, I'll adjust PR)

**Comment by vjuranek** [comment]
Updated again (removed tests related to ISPN-8429 as it's not a bug)

File: core/src/test/java/org/infinispan/tx/totalorder/statetransfer/TotalOrderStateTransferFunctionalTest.java [hunk 1]
```diff
@@ -15,26 +16,32 @@
...
    @Override
    public Object[] factory() {
       return new Object[] {
-         new TotalOrderStateTransferFunctionalTest("dist-to-1pc-nbst", CacheMode.DIST_SYNC, false, false),
-         new TotalOrderStateTransferFunctionalTest("dist-to-2pc-nbst", CacheMode.DIST_SYNC, true, false),
-         new TotalOrderStateTransferFunctionalTest("repl-to-1pc-nbst", CacheMode.REPL_SYNC, true, true),
-         new TotalOrderStateTransferFunctionalTest("repl-to-2pc-nbst", CacheMode.REPL_SYNC, true, false),
+         new TotalOrderStateTransferFunctionalTest("dist-to-1pc-nbst", CacheMode.DIST_SYNC, false, false, StorageType.OBJECT),
+         new TotalOrderStateTransferFunctionalTest("dist-to-2pc-nbst", CacheMode.DIST_SYNC, true, false, StorageType.OBJECT),
+         new TotalOrderStateTransferFunctionalTest("dist-to-1pc-nbst-off-heap", CacheMode.DIST_SYNC, false, false, StorageType.OBJECT),
```
Comment (wburns) [inline core/src/test/java/org/infinispan/tx/totalorder/statetransfer/TotalOrderStateTransferFunctionalTest.java]: Think you meant to put `StorageType.OFF_HEAP` for a few of these

File: core/src/test/java/org/infinispan/tx/ContextAffectsTransactionRepeatableReadTest.java [hunk 2]
```diff
@@ -15,6 +17,15 @@
  */
 @Test (groups = "functional", testName = "tx.ContextAffectsTransactionRepeatableReadTest")
 public class ContextAffectsTransactionRepeatableReadTest extends ContextAffectsTransactionReadCommittedTest {
+
+   @Factory
+   @Override
+   public Object[] factory() {
+      return new Object[] {
+            new ContextAffectsTransactionRepeatableReadTest().withStorage(StorageType.OBJECT)
```
Comment (wburns) [inline core/src/test/java/org/infinispan/tx/ContextAffectsTransactionRepeatableReadTest.java]: So this test doesn't work with off heap? If not maybe we should have a comment.

File: core/src/test/java/org/infinispan/tx/ContextAffectsTransactionReadCommittedTest.java [hunk 3]
```diff
@@ -27,9 +29,26 @@
  */
 @Test (groups = "functional", testName = "tx.ContextAffectsTransactionReadCommittedTest")
 public class ContextAffectsTransactionReadCommittedTest extends SingleCacheManagerTest {
+
+   protected StorageType storage;
+
+   @Factory
+   public Object[] factory() {
+      return new Object[] {
+            new ContextAffectsTransactionReadCommittedTest().withStorage(StorageType.OBJECT),
```
Comment (wburns) [inline core/src/test/java/org/infinispan/tx/ContextAffectsTransactionReadCommittedTest.java]: Just for completeness, we should add BINARY.

File: core/src/test/java/org/infinispan/tx/TerminatedCacheWhileInTxTest.java [hunk 4]
```diff
@@ -38,13 +40,29 @@
...
+            new TerminatedCacheWhileInTxTest().withStorage(StorageType.OBJECT),
+            new TerminatedCacheWhileInTxTest().withStorage(StorageType.OFF_HEAP)
+      };
+   }
+
    @Override
    protected EmbeddedCacheManager createCacheManager() throws Exception {
       ConfigurationBuilder c = TestCacheManagerFactory.getDefaultCacheConfiguration(true);
       c.transaction().cacheStopTimeout(10000);
+      c.memory().storageType(StorageType.OFF_HEAP);
```
Comment (wburns) [inline core/src/test/java/org/infinispan/tx/TerminatedCacheWhileInTxTest.java]: Think this was supposed to be `storage`.

File: core/src/test/java/org/infinispan/tx/synchronisation/TransactionsSpanningCachesSyncTest.java [hunk 5]
```diff
@@ -14,14 +16,22 @@
 @Test(groups = "functional", testName = "tx.synchronisation.TransactionsSpanningCachesSyncTest")
 public class TransactionsSpanningCachesSyncTest extends TransactionsSpanningCaches {
 
+   @Factory
+   @Override
+   public Object[] factory() {
+      return new Object[] {
+            new TransactionsSpanningCachesSyncTest().withStorage(StorageType.OBJECT, StorageType.OBJECT)
```
Comment (wburns) [inline core/src/test/java/org/infinispan/tx/synchronisation/TransactionsSpanningCachesSyncTest.java]: This one doesn't work with off heap either? Maybe just add a comment.

**Comment by vjuranek** [comment]
PR updated. Thanks for spotting several mistakes - everything works also with off-heap (I didn't find any issue in TX related to off-heap, except known limitations), I just forgot to add back off-heap after playing with the tests.

**Comment by wburns** [comment]
Integrated into master, thanks @vjuranek !

**Comment by gustavonalle** [comment]
I have some errors when running in my local env. Haven't seen them on CI, maybe because it's slower?
https://issues.jboss.org/browse/ISPN-8547

**Comment by wburns** [comment]
That should have been fixed in https://github.com/infinispan/infinispan/commit/dace20e9f6c4366f51f89c96c4fac0ed7d2e05ca Are you on the latest master?

**Comment by gustavonalle** [comment]
@wburns Indeed it seems related, but I believe I am up-to-date:
```bash  
$ git log --oneline | head -n 1
d5ff235875 ISPN-8443 Remove entries from all stored on entry removal
```

**Pull request merged**